test/fuzz/tcti/tcti-fuzzing.c

```diff
@@ -0,0 +1,274 @@
+/* SPDX-License-Identifier: BSD-2 */
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * All rights reserved.
+ */
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <sys/time.h>
+#include <inttypes.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include "tss2_mu.h"
+#include "tss2_tcti_fuzzing.h"
+
+#include "tcti-fuzzing.h"
+#include "tss2-tcti/tcti-common.h"
+#include "util/key-value-parse.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * Using the data and size fields of the fuzzing TCTI memcpy the data into the
+ * structures given via va_list. Caller will pass the sysContext and number of
+ * arguments that follow. Following the count of arguments caller should pass
+ * the sizeof the next argument, which shall be a pointer to the structure to be
+ * filled.
+ *
+ * Example:
+ *
+ *     TPMI_DH_OBJECT keyA = {0};
+ *     TPM2B_ECC_POINT inQsB = {0};
+ *     TPM2B_ECC_POINT inQeB = {0};
+ *     TPMI_ECC_KEY_EXCHANGE inScheme = {0};
+ *     UINT16 counter = {0};
+ *
+ *     ret = fuzz_fill (
+ *         sysContext,
+ *         10,
+ *         sizeof (keyA), &keyA,
+ *         sizeof (inQsB), &inQsB,
+ *         sizeof (inQeB), &inQeB,
+ *         sizeof (inScheme), &inScheme,
+ *         sizeof (counter), &counter
+ *     );
+ *     if (ret) {
+ *         ... handle failure
+ *     }
+ */
+int
+fuzz_fill (
+        TSS2_SYS_CONTEXT *sysContext,
+        size_t count,
+        ...)
+{
+    va_list ap;
+    const uint8_t *data = NULL;
+    const uint8_t *curr = NULL;
+    size_t size = 0U;
+    size_t i = 0U;
+    void *dest;
+    size_t length = 0U;
+    size_t combined = 0U;
+    _TSS2_SYS_CONTEXT_BLOB *ctx = NULL;
+    TSS2_TCTI_FUZZING_CONTEXT *tcti_fuzzing = NULL;
+
+    ctx = syscontext_cast (sysContext);
+    tcti_fuzzing = tcti_fuzzing_context_cast (ctx->tctiContext);
+    data = tcti_fuzzing->data;
+    size = tcti_fuzzing->size;
+
+    va_start (ap, count);
+
+    for (i = 0U; i < (count / 2); ++i) {
+        length = va_arg (ap, size_t);
+        dest = va_arg (ap, void *);
+        if (size > (combined + length)) {
+            curr = &data[combined];
```

Comment:
I think that be just:
curr = data;

va_arg() performs the iteration. If one would do something like:

TPMI_DH_OBJECT A = {0};
TPMI_DH_OBJECT B = {0};

fuzz_fill (sysContext, 4, sizeof (B), &B, sizeof(A), &A));

then the &data[combined] in the second iteration would point somewhere after the object B.

Reply:
Note after in person review: John needs to change the names here